napari/layers/surface/_tests/test_surface.py

```diff
@@ -214,6 +214,51 @@ def test_shading():
     assert layer.shading == shading
 
 
+def test_texture():
+    """Test setting texture"""
+    np.random.seed(0)
+    vertices = np.random.random((10, 3))
+    faces = np.random.randint(10, size=(6, 3))
+    values = np.random.random(10)
+    data = (vertices, faces, values)
+
+    texture = np.random.random((32, 32, 3)).astype(np.float32)
+    texcoords = vertices[:, :2]
+    layer = Surface(data, texture=texture, texcoords=texcoords)
+
+    np.testing.assert_allclose(layer.texture, texture)
+    np.testing.assert_allclose(layer.texcoords, texcoords)
+    assert layer._has_texture
+
+    layer.texture, layer.texcoords = None, texcoords
+    assert not layer._has_texture
+
+    layer.texture, layer.texcoords = texture, None
+    assert not layer._has_texture
+
+    layer.texture, layer.texcoords = None, None
+    assert not layer._has_texture
+
+    layer.texture, layer.texcoords = texture, texcoords
+    assert layer._has_texture
```

Comment:
this is sort of exceeding my Python knowledge: does `a, b = c, d` expand to `a = c; b = d` in bytecode? Does this mean that it's fine for users to update one, then the other in code, in any order? Also, is it possible to have a texture incompatible with texcoords or are texcoords normalised to [0, 1] in the texture size?

Reply:
>  does a, b = c, d expand to a = c; b = d in bytecode?

In the spirit yes, in practice not exactly in case of of the item in on LHS and RHS.

```
In [8]: x.a = 0
set a

In [9]: x.b = 0
set b

In [10]: x.a, x.b = 1, x.a
set a
set b



In [11]: print(x.a, x.b)
1 0 
```
`x.b` is 0, not 1, ; `x.a = 1; x.b = x.a` would give `1 1`

More specifically:


```
In [13]: dis.dis('x.a, x.b = 1, x.a')
  0           0 RESUME                   0

  1           2 LOAD_CONST               0 (1)
              4 LOAD_NAME                0 (x)
              6 LOAD_ATTR                1 (a)
             16 SWAP                     2
             18 LOAD_NAME                0 (x)
             20 STORE_ATTR               1 (a)
             30 LOAD_NAME                0 (x)
             32 STORE_ATTR               2 (b)
             42 LOAD_CONST               1 (None)
             44 RETURN_VALUE

In [14]: dis.dis('a,b = c,d')
  0           0 RESUME                   0

  1           2 LOAD_NAME                0 (c)
              4 LOAD_NAME                1 (d)
              6 SWAP                     2
              8 STORE_NAME               2 (a)
             10 STORE_NAME               3 (b)
             12 LOAD_CONST               0 (None)
             14 RETURN_VALUE
```

But here it does not matter much.

Reply:
Thanks for the deep explanation. I admit I wasn't thinking about it at the bytecode level here. This just felt like a clear way to write the test for each combination.

> Does this mean that it's fine for users to update one, then the other in code, in any order?

Right I think it shouldn't matter, other than if you update one you might briefly see the wrong texture or mapping if the coords change.

Reply:
Yeah, I just wanted to make sure that users didn't *have* to use an expression like this one and could just update these things one by one normally. 😊